bayes_kit/drghmc.py

```diff
@@ -0,0 +1,470 @@
+
+        Returns:
+            Iterator of draws from DrGhmcDiag sampler
+        """
+        return self
+
+    def __next__(self) -> DrawAndLogP:
+        """Return the next draw from this sampler, along with its log density.
+
+        Yields:
+            Tuple of (draw, log density)
+        """
+        return self.sample()
+
+    def joint_logp(self, theta: VectorType, rho: VectorType) -> float:
+        """Log density of draw (theta, rho) under the unnormalized Gibbs pdf.
+
+        The Gibbs distribution, also known as the canonical or Boltzmann distribution,
+        is a probability density function (pdf) that depends on the Hamiltonian. The
+        Hamiltonian is the sum of the potential and kinetic energies, defined by the
+        position and momentum respectively.
+
+        Args:
+            theta: position (model parameters)
+            rho: momentum
+
+        Returns:
+            unnormalized, joint log density of draw (theta, rho)
+        """
+
+        if not self._cache:
+            logp, grad = self._model.log_density_gradient(theta)
+            self._cache = [(logp, np.asanyarray(grad))]
+        else:
+            logp, _ = self._cache[-1]
+
+        potential = -logp
+        kinetic: float = 0.5 * np.dot(rho, self._metric * rho)
+        hamiltonian = potential + kinetic
+        return -hamiltonian
+
+    def leapfrog(
+        self,
+        theta: VectorType,
+        rho: VectorType,
+        stepsize: float,
+        stepcount: int,
+    ) -> tuple[VectorType, VectorType]:
+        """Simulate Hamiltonian dynamics by leapfrog integration for draw (theta, rho).
+
+        Discretize and solve Hamilton's equations using the leapfrog integrator with
+        specified stepsize and stepcount.
+
+        When leapfrog integration is followed by a momentum flip, this generates a new,
+        proposed draw (theta_prop, rho_prop) from the current draw (theta, rho).
+
+        Args:
+            theta: position (model parameters)
+            rho: momentum
+            stepsize: stepsize in each leapfrog step
+            stepcount: number of leapfrog steps
+
+        Returns:
+            Hamiltonian dynamics simulated for draw (theta, rho)
+        """
+        theta = np.array(theta, copy=True)  # copy b/c numpy's += mutates original array
+        grad: ArrayLike  # mypy infers too strict a type when reading from cache
+
+        logp, grad = self._cache[-1]
+        rho_mid = rho + 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        theta += stepsize * rho_mid
+
+        for _ in range(stepcount - 1):
+            logp, grad = self._model.log_density_gradient(theta)
+            rho_mid += stepsize * np.multiply(self._metric, grad).squeeze()
+            theta += stepsize * rho_mid
+
+        logp, grad = self._model.log_density_gradient(theta)
+        rho = rho_mid + 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+
+        self._cache.append((logp, np.asanyarray(grad)))
```

Comment:
I can't find where we benefit from keeping a list of cached values.   What's the motivation behind appending cached values instead of only maintaining the last `logp` and `grad`?

Reply:
When we have multiple proposal attempts and recursively defined ghost proposals, we actually need to track multiple `grad` values in a cache -- simply tracking the most recent `grad` value is not enough.

Reply:
I definitely did not catch that the first time around :) Thanks!